napari/util/keybindings.py

```diff
@@ -0,0 +1,582 @@
+
+    return '-'.join(modifiers + (key,))
+
+
+def normalize_key_sequence(seq):
+    """Normalize key sequence to make it easily comparable.
+
+    All aliases are converted and modifier orders are fixed to:
+    Control, Alt, Shift, Meta
+
+    Parameters
+    ----------
+    seq : str
+        Key sequence.
+
+    Shift pressed with any letter will be consumed
+    to transform it to upper case, e.g. Shift + w => W.
+    Otherwise, the letter will be lower case.
+    Shift will also be consumed without transformation when pressed
+    with any other non-special key unless Control is also a modifier.
+
+    Returns
+    -------
+    normalized_seq : str
+        Normalized key sequence.
+    """
+    seq = _expand_shorthand(seq)
+
+    key, modifiers = parse_seq(seq)
+
+    if len(key) != 1 and key not in SPECIAL_KEYS:
+        raise TypeError(f'invalid key {key}')
+
+    for modifier in modifiers:
+        if modifier not in MODIFIER_KEYS.values():
+            raise TypeError(f'invalid modifier key {modifier}')
+
+    return components_to_seq(key, modifiers)
+
+
+def _determine_binding_op(arg):
+    """Determine the binding operation based on the argument passed.
+
+    None     -> BindOperation.UNBIND
+    str      -> BindOperation.REBIND
+    callable -> BindOperation.BIND
+    """
+    if arg is None:
+        return BindOperation.UNBIND
+    elif isinstance(arg, str):
+        return BindOperation.REBIND
+    elif isinstance(arg, Callable):
+        return BindOperation.BIND
+    raise TypeError('expected arg to be a string, callable, or None; '
+                    f'got {type(arg)}')
+
+
+def _bind_key(keybindings, seq, func):
+    """Bind a key sequence to a keymap.
+    """
+    keybindings[seq] = func
+
+
+def _unbind_key(keybindings, seq):
+    """Unbind a function from a keymap based on its sequence and return it.
+    """
+    return keybindings.pop(seq)
+
+
+def _rebind_key(keybindings, new_seq, orig_seq):
+    """Rebind a function from one key sequence to another.
+    """
+    func = _unbind_key(keybindings, orig_seq)
+    _bind_key(keybindings, new_seq, func)
+
+
+def bind_key(keybindings, seq, func):
+    """Bind a key sequence to a keymap.
+
+    Parameters
+    ----------
+    keybindings : dict of str: callable
+        Keymap to modify.
+    seq : str
+        Key sequence.
+    func : callable, str, or None
+        Callable to bind to the key sequence.
+        If a string is passed, instead perform a rebind operation.
+        If ``None`` is passed, unbind instead.
+    """
+    seq = normalize_key_sequence(seq)
+
+    bind_op = _determine_binding_op(func)
+
+    if bind_op == BindOperation.UNBIND:
+        try:
+            _unbind_key(keybindings, seq)
+        except KeyError:
+            pass
+    elif bind_op == BindOperation.REBIND:
+        orig_seq = normalize_key_sequence(func)
+        _rebind_key(keybindings, seq, orig_seq)
+    else:
+        _bind_key(keybindings, seq, func)
```

Comment:
I don't love `_rebind`... Can we do without it?

I'm also not sure that `bind_key` should ever unbind, but I can see how `bind_key(dict, 'r', None)` might be an intuitive way to deal with this. BUT, not with `unbind_key` around! There should be one way to do this, I think.

Reply:
@jni does that mean that if you try and bind to an already bound key the effect would be additive, not to replace the existing binding? I'm not sure what is most intuitive, but best practice would probably be to let the user know somehow that there is already something there

Reply:
I'm happy for `bind` to always replace, OR for it to be additive (with a warning) and having to do unbind/bind in sequence. I just think we should keep this API small. Maybe even having just `bind(..., None)` being `unbind`, then the API is really just one function.